libs/language-server/src/lib/ast/wrappers/pipeline-wrapper.ts

```diff
@@ -0,0 +1,83 @@
+// SPDX-FileCopyrightText: 2023 Friedrich-Alexander-Universitat Erlangen-Nurnberg
+//
+// SPDX-License-Identifier: AGPL-3.0-only
+
+import {
+  BlockDefinition,
+  CompositeBlocktypeDefinition,
+  PipelineDefinition,
+} from '../generated/ast';
+
+import { AstNodeWrapper } from './ast-node-wrapper';
+import { PipeWrapper, createWrappersFromPipeChain } from './pipe-wrapper';
+
+export class PipelineWrapper
+  implements AstNodeWrapper<PipelineDefinition | CompositeBlocktypeDefinition>
+{
+  public readonly astNode: PipelineDefinition | CompositeBlocktypeDefinition;
+
+  allPipes: PipeWrapper[] = [];
+
+  constructor(
+    pipelineDefinition: PipelineDefinition | CompositeBlocktypeDefinition,
+  ) {
+    this.astNode = pipelineDefinition;
+
+    this.allPipes = pipelineDefinition.pipes.flatMap((pipe) =>
+      createWrappersFromPipeChain(pipe),
+    );
+  }
+
+  static canBeWrapped(pipelineDefinition: PipelineDefinition): boolean {
+    for (const pipeDefinition of pipelineDefinition.pipes) {
+      for (
+        let chainIndex = 0;
+        chainIndex < pipeDefinition.blocks.length - 1;
+        ++chainIndex
+      ) {
+        if (!PipeWrapper.canBeWrapped(pipeDefinition, chainIndex)) {
+          return false;
+        }
+      }
+    }
+    return true;
+  }
+
+  getStartingBlockPipes(): PipeWrapper[] {
+    return this.allPipes.filter((pipe) => {
+      const fromBlock = pipe.from;
+      const isToOfOtherPipe =
+        this.allPipes.filter((p) => p.to === fromBlock).length > 0;
+      return !isToOfOtherPipe;
+    });
+  }
+
+  getStartingBlocks(): BlockDefinition[] {
+    const startingBlocks = this.getStartingBlockPipes().map((p) => p.from);
+
+    // Special case: the extractor is reused for multiple paths
+    // Thus, we remove duplicates
+    const withoutDuplicates = [...new Set(startingBlocks)];
+    return withoutDuplicates;
+  }
+
+  getOutgoingPipes(blockDefinition: BlockDefinition): PipeWrapper[] {
+    return this.allPipes.filter((pipe) => {
+      return pipe.from === blockDefinition;
+    });
+  }
+
+  followOutgoingPipes(blockDefinition: BlockDefinition): BlockDefinition[] {
```

Comment:
I strongly feel we should follow terminology of directed acyclical graphs here.

This would mean for a block X:

- blocks reachable by following an outgoing pipe from X are called `children`
- blocks reachable by following an incoming pipe from X are called `parents`
- if you need to follow more than one pipe from X as described above the blocks are called `ancestors` or `descendants`

So for these functions it would mean renaming according to that:
- `followOutgoingPipes` would become `getChildren` (`getChildrenBlocks` works as well for me)
- `followIngoingPipes` would become `getParents` (`getParentBlocks`)

Reply:
This extends to variablenaming, e.g. `fromBlock` in line 48 would become `parentBlock`. I think this is a) much clearer and b) makes it much easier to read and understand what is going on here with the DAG image in the back of the mind.

Reply:
Addressed in [78bba05](https://github.com/jvalue/jayvee/pull/491/commits/78bba058e40589972164b0a7725a64fc60ba153a)